keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -532,44 +530,47 @@ reconstructed by the driver. This has no effect on already allocated claims
 because the allocation result is tracked in those claims, not the
 `ResourceSlice` objects (see [below](#state-and-communication)).
 
-#### Devices as a named list of attributes
+#### Devices as a named list
 
-Embedded inside each `ResourceSlice` is a list of one or more devices, represented as a named list of attributes.
+Embedded inside each `ResourceSlice` is a list of one or more devices, each with sets of attributes
+and quantities.
 
 ```yaml
 kind: ResourceSlice
 apiVersion: resource.k8s.io/v1alpha3
 ...
 spec:
   # The node name indicates the node.
+  #
   # Each driver on a node provides pools of devices for allocation,
-  # with unique device names inside each pool.
-  # Usually, but not necessarily, that pool name is the same as the
-  # node name.
+  # with unique device names inside each pool. A pool might be defined
+  # by more than one ResourceSlice.
   nodeName: worker-1
-  poolName: worker-1
-  driverName: cards.dra.example.com
+  pool:
+    ...
+  driver: cards.dra.example.com
   devices:
   - name: card-1 # unique inside the worker-1 pool
-    attributes:
-    - name: manufacturer # a vendor-specific attribute, automatically qualified as cards.dra.example.com/manufacturer
-      string: Cards-are-us Inc.
-    - name: productName:
-      string: ACME T1000 16GB
-    - name: driverVersion
-      version: 1.2.3
-    - name: runtimeVersion
-      string: 11.1.42
-    - name: memory
-      quantity: 16Gi
-    - name: powerSavingSupported
-      bool: true
-    - name: dra.k8s.io/pciRoot # a fictional standardized attribute, not actually part of this KEP
-      string: pci-root-0
+    basic:
+      attributes:
+        manufacturer: # a vendor-specific attribute, automatically qualified as cards.dra.example.com/manufacturer
+          string: Cards-are-us Inc.
+        productName:
+          string: ACME T1000 16GB
+        driverVersion:
+          version: 1.2.3
+        runtimeVersion:
+          version: 11.1.42
+        powerSavingSupported:
+          bool: true
+        dra.k8s.io/pciRoot: # a fictional standardized attribute, not actually part of this KEP
+          string: pci-root-0
+      capacity:
+        memory: 16Gi
 ```
 
-Compared to labels, attributes have values of exactly one type. As
-described later on, these attributes can be used in CEL expressions to select a
+Compared to labels, attributes have values of exactly one type. Quantities are defined
+in the separate `capacity` map. As described later on, both sets can be used in CEL expressions to select a
```

Comment:
Can we add some more text here explaining why there is a separation of Attributes / Capacities?

Reply:
As far as I am concerned, that is "because someone asked me to". It never made a lot of sense to me.

Reply:
I believe the rationale is so that we can (in the future) reference these outside of CEL expressions as explicit resource requests, i.e. instead of:
```
      selectors:
      - cel:
          expression: "device.capacity['memory'].compareTo(quantity('80Gi')) >= 0"
```
you could do:
```
      selectors:
      - capacity:
          memory: 80GB
```
This doesn't make sense for arbitrary attributes, but it could for quantities.

Reply:
I still don't quite buy that argument. Such a "capacity" selector could also be called "requirements" with the implied semantic of "at least this" and then it would make sense for quantity and version attributes:
```
selectors:
- requirements:
    memory: 80GB
    driverVersion: 2.0.0
```

We talked about "consumable quantities", which would be a different semantic. Because the semantic of those is different, we can't use the existing `BasicDevice` and would have to define something new. 

I'm pretty sure `BasicDevice.Capacity` will always have the same semantic as it has now.

Reply:
I am unclear why we could not support consumption in the existing model?

Reply:
The scheduler cannot distinguish between "this capacity A is consumable" and "that capacity B is not consumable". The semantic is always the latter right now.

We need a new API for marking capacities as consumable, and I don't see how to fit that into the `BasicDevice` struct.

Reply:
As I remember it, there was a dislike for having an attribute that looks like a consumable capacity elsewhere.

Reply:
I don't object to Quantity-typed attributes AS LONG AS we can say why we need them to NOT be consumable.

> Current GPU memory is not consumable

Only because we do not have a "request" consumption model.  Consumption is how we indicate that it is unavailable to others - once we have sub-device capabilities, this will be very important, right?

Reply:
> Only because we do not have a "request" consumption model.

The hardware also has to support it. We could define such a model, but who is going to use it for what?

MIG has sub-partitions, but only with pre-defined memory, which then is again an attribute of that sub-partition, not some consumable capacity. @klueska: is that correct?

>  once we have sub-device capabilities, this will be very important, right?

Once we add that, we need API extensions. We cannot simply re-intepret an existing field.

IMHO the right approach would be to add a "Quantity" type to atttributes and to remove the `Capacity` field until we define it in a way which is not "another attribute".

I don't care whether we then keep or remove the "Integer" attribute type.

Reply:
I thought there was at least one non-MIG mode which allowed device sharing - https://developer.nvidia.com/blog/improving-gpu-utilization-in-kubernetes/#cuda_multi-process_service

> Once we add that, we need API extensions

Do we have the request model worked out yet?  What part of it doesn't work?

Reply:
You don't want to be doing MPS sharing at this level.

There are a bunch of gotchas:
* The same Linux user needs to be used to run on each MPS "partition" for a given GPU. 
* There is no fault isolation between different MPS partitions (meaning that if one container's job borks the GPU, all MPS processes on that GPU go down as well).
* A container can escape out of MPS isolation at anytime by simply unsetting a few envvars before launching it's job. It then has free rein to do what it wants on the full GPU.

The best (and only really correct) way to do MPS is via what I have been calling "user-mediated" sharing.

The idea being that the user asks for a full GPU with MPS sharing mode enabled (via opaque parameters), and then the user takes charge to carve up the GPU as he sees fit between his workloads (via `CUDA_MPS_PINNED_DEVICE_MEM_LIMIT` and `CUDA_MPS_ACTIVE_THREAD_PERCENTAGE ` envars set in each container referencing the shared GPU).

This is the reason we've always been hesitant to add support for MPS into our standard device plugin. I wouldn't recommend adding support for MPS via a request model if we had one.

I only see NVIDIA GPUs making use of a request model if/when linux support for GPU cgroups ever materializes. That way we can set limits / percentages on GPU memory and GPU compute the same way we do CPU memory and CPU compute.  Last I heard this effort was stalled though, and I'm not sure what the current status of it is.

Reply:
So how do we proceed?

I see the following options:
- Keep it-as is (= semantically it is an attribute), promote the current API to beta.
- Remove `BasicDevice.Capacity` and replace it with a `Quantity` attribute type in v1beta1. This implies removing v1alpha3 in that same release or doing manual conversion with a slight hack: when converting v1alpha3 -> v1beta1, we have to assume that attribute and capacity names don't clash. That's not required at the moment in v1alpha3, but most likely true to all existing DRA drivers.

cc @johnbelamaric 

I'm keeping the PR open because I don't want to loose this discussion. We could merge without adding further clarifications to the README.md in this PR, that can come later when describing next steps.

Reply:
Per-device consumable capacities don't work well right now with current NVIDIA technology. There are use cases for it in things like Pod offloading to DPUs, but those are not our current focus. So, we have deprioritized building the device model support for that as of today. I think there is reason to believe this will be a useful thing eventually, maybe for GPUs, but maybe not.

> The intent was that all capacity is consumable, or else it is not in capacity.
Agreed.

> The hardware also has to support it. 
I actually disagree with this. Ideally the hardware supports it, but there is still value allowing the scheduler to dole out partial resources among cooperative users, even if the cooperation is no enforced. The problem is that there are more severe gotchas today with NVIDIA, at least, as Kevin describes. A noisy neighbor being rude and starving other processes is not nearly so bad as one that can read their memory or crash them.

In any case, I agree with Tim that it is the *consumption* semantic that would define whether we consider partial allocations of the consumable capacities. So, all "capacity" ones are consumable; we just haven't built the *claim* mechanism for that yet. In CEL, the reference to the "foo" capacity is the raw capacity, regardless of allocations, and *that* semantic will not change.

So - to me, we should keep it as is and advance it to beta.

Reply:
> So, all "capacity" ones are consumable

@klueska already refuted that for the NVidia DRA driver. It publishes "memory" as a capacity because it has to (no quantity attribute), but if we defined and implemented some mechanism in the scheduler which then grants access to that same `BasicDevice` in two unrelated `ResourceClaims`, it won't necessarily work (needs to be the same Linux user) and won't be safe (fault isolation, limit enforcement).

I can see that if all users of such a BasicDevice agree, then Kubernetes might let them do it anyway, but I am not sure whether that's wise. It feels like a footgun to me.

> So - to me, we should keep it as is and advance it to beta.

I think DRA drivers should opt into letting users consume capacity. If they don't opt in, it's not consumable. Another argument is that we currently might not provide enough information to Kubernetes about how it can split up a capacity if it is consumable (the block size, for example).

I'd prefer to have `BasicDevice.Capacity` as a `map[QualifiedName]DeviceCapacty` with
```
type DeviceCapacity struct {
    Quantity resource.Quantity
    <future fields to be added here for consumable capacity>
}
```

That won't round-trip easily into v1alpha3 where `Capacity` is a `map[QualifiedName]resource.Quantity`. But let's not let ourselves be constrained by that. The priority has to be to have a sane `v1beta1` API. Whether we want to and can continue to support `v1alpha3` in 1.32 is less important.

I've created https://github.com/kubernetes/kubernetes/issues/127386 to track this questions. Let's continue there.